contracts/src/set_verifier.rs

```diff
@@ -116,4 +133,110 @@ where
 
         Ok((image_id, image_url))
     }
+
+    /// Returns the seal if of the given verified root.
+    pub async fn get_verified_root_seal(&self, root: B256) -> Result<Bytes> {
+        self.query_verified_root_event(root, None, None).await
+    }
+
+    async fn get_latest_block(&self) -> Result<u64> {
+        self.instance
+            .provider()
+            .get_block_number()
+            .await
+            .context("Failed to get latest block number")
+    }
+
+    /// Query the VerifiedRoot event based on the root and block options.
+    /// For each iteration, we query a range of blocks.
+    /// If the event is not found, we move the range down and repeat until we find the event.
+    /// If the event is not found after the configured max iterations, we return an error.
+    /// The default range is set to 100 blocks for each iteration, and the default maximum number of
+    /// iterations is 100. This means that the search will cover a maximum of 10,000 blocks.
+    /// Optionally, you can specify a lower and upper bound to limit the search range.
+    async fn query_verified_root_event(
+        &self,
+        root: B256,
+        lower_bound: Option<u64>,
+        upper_bound: Option<u64>,
+    ) -> Result<Bytes> {
+        let mut upper_block = upper_bound.unwrap_or(self.get_latest_block().await?);
+        let start_block = lower_bound.unwrap_or(upper_block.saturating_sub(
+            self.event_query_config.block_range * self.event_query_config.max_iterations,
+        ));
+
+        // Loop to progressively search through blocks
+        for _ in 0..self.event_query_config.max_iterations {
+            // If the current end block is less than or equal to the starting block, stop searching
+            if upper_block <= start_block {
+                break;
+            }
```

Comment:
We should search the single block if they are equal, no? Or is upper-bound not inclusive. If it is not, then it seems like the default above should be latest + 1
```suggestion
            // If the current end block is less than to the starting block, stop searching
            if upper_block < start_block {
                break;
            }
```